### PR TITLE
Merge Java branch

### PR DESCRIPTION
**Pull request by jogo**
The java branch adds support for some java stacktraces along with several related improvements. See individual commits for more details.

Addresses #1

### CODE REVIEW

File: git_stacktrace/git.py [hunk 1]
```diff
@@ -9,6 +9,37 @@
...
 
 
+class GitFile(object):
+    """Track filename and if file was added/removed or modified."""
+    ADDED = 'A'
+    DELETED = 'D'
+    MODIFIED = 'M'
+    COPY_EDIT = 'C'
+    RENAME_EDIT = 'R'
+    VALID = [ADDED, DELETED, MODIFIED, COPY_EDIT, RENAME_EDIT]
```
Comment (jparise) [inline git_stacktrace/git.py]: `frozenset` instead?

File: git_stacktrace/git.py [hunk 2]
```diff
@@ -9,6 +9,37 @@
...
+        self.filename = filename
+        if state not in GitFile.VALID:
+            raise Exception("Invalid git file state: %s" % state)
+        self.state = state
+
+    def __repr__(self):
+        return self.filename
+
+    def __cmp__(self, other):
+        if type(other) == str or type(other) == unicode:
```
Comment (jparise) [inline git_stacktrace/git.py]: Why not `isinstance(other, basestring)`?

File: git_stacktrace/parse_trace.py [hunk 3]
```diff
@@ -46,9 +52,41 @@ def extract_python_traceback(self, blob):
...
+
+class PythonTraceback(Traceback):
+    """Parse Traceback string."""
+
+    def extract_traceback(self, lines):
+        """Convert traceback string into a traceback.extract_tb format"""
         # TODO better logging if cannot read traceback
         # filter out traceback lines
+        self.header = lines[0] + '\n'
+        if len(lines[-1]) > 0 and not lines[-1].startswith(' '):
```
Comment (jparise) [inline git_stacktrace/parse_trace.py]: `len(lines[-1]) > 0` can be just `lines[-1]`

File: git_stacktrace/parse_trace.py [hunk 4]
```diff
@@ -57,32 +95,100 @@ def extract_python_traceback(self, blob):
...
-        if lines != new_lines:
-            raise Exception("Incorrectly extracted traceback information")
-
-    def filter_site_packages(self):
-        filtered = []
-        for line in self.lines:
-            if '/site-packages/' not in line.trace_filename:
-                filtered.append(line)
-        self.lines = filtered
+        if lines != new_lines or len(self.lines) == 0:
```
Comment (jparise) [inline git_stacktrace/parse_trace.py]: `or not self.lines`

File: git_stacktrace/parse_trace.py [hunk 5]
```diff
@@ -57,32 +95,100 @@ def extract_python_traceback(self, blob):
...
+
+    def extract_traceback(self, lines):
+        if not lines[0].startswith('\t'):
+            self.header = lines[0] + '\n'
+        lines = [line for line in lines if line.startswith('\t')]
+        extracted = []
+        for line in lines:
+            extracted.append(self._extract_line(line))
+        self.lines = extracted
+        if len(self.lines) == 0:
```
Comment (jparise) [inline git_stacktrace/parse_trace.py]: `if not self.lines:`

File: git_stacktrace/parse_trace.py [hunk 6]
```diff
@@ -57,32 +95,100 @@ def extract_python_traceback(self, blob):
...
+        filename = split[-1]
+        if line.native_method:
+            return "\tat %s.%s.%s(Native Method)\n" % (path, line.class_name, line.function_name)
+        return "\tat %s.%s.%s(%s:%d)\n" % (path, line.class_name, line.function_name, filename, line.line_number)
+
+    def format_lines(self):
+        result = ''
+        for line in self.lines:
+            result += self._format_line(line)
+        return result
```
Comment (jparise) [inline git_stacktrace/parse_trace.py]: Can this be just `''.join(map(self._format_line, self.lines))`?

File: git_stacktrace/result.py [hunk 7]
```diff
@@ -6,23 +6,41 @@ class Result(object):
...
+    def add_file(self, git_file):
+        if git_file.state in [git.GitFile.ADDED, git.GitFile.COPY_EDIT]:
+            self.files_added.add(git_file.filename)
+        elif git_file.state in [git.GitFile.DELETED]:
+            self.files_deleted.add(git_file.filename)
+        else:
+            self.files_modified.add(git_file.filename)
+
     def __hash__(self):
         return self.commit.__hash__()
```
Comment (jparise) [inline git_stacktrace/result.py]: Or: `return hash(self.common)`

File: git_stacktrace/parse_trace.py (hunk 6 above)
Comment (jogo) [inline git_stacktrace/parse_trace.py]: Yes, good idea. I actually personally find the map notation easier to read.

File: git_stacktrace/result.py (hunk 7 above)
Comment (jogo) [inline git_stacktrace/result.py]: good idea

**Pull request merged**